pkg/kubelet/prober/prober_manager.go

```diff
@@ -48,6 +48,21 @@ var ProberResults = metrics.NewCounterVec(
 		"pod_uid"},
 )
 
+var ProberDuration = metrics.NewHistogramVec(
+	&metrics.HistogramOpts{
+		Subsystem:      "prober",
+		Name:           "probe_duration",
+		Help:           "Duration in seconds for a probe response",
+		StabilityLevel: metrics.ALPHA,
+	},
+	[]string{"probe_type",
+		"result",
+		"container",
+		"pod",
+		"namespace",
+		"pod_uid"},
```

Comment:
All these labels are very high cardinality and go against our [instrumentation guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md#dimensionality--cardinality). Reason being that for this histogram you'll have the following number of timeseries generated:
```
16(histogram) * M_containers * O_pods * P_namespaces * 3(probe_type) * 3(results)
```
Which turns out to be pretty insane and can very easily cause the monitoring platform to blow up. As such I would advise removing these labels. But that might not make the metric meaningful anymore so is there any other information that you could use to differentiate probes besides their types?

Reply:
As far as I can tell all of these labels are required to define uniqueness w/ respect to the provenance of the probe. Can "pod_uid" be eliminated?

Any thoughts here @SergeyKanzhelev ?

Reply:
Having both `pod` and `pod_uid` feels a bit redundant, but it does not add more cardinality to the metric since `pod_uid` already has a unique value per pod. So eliminating the `pod_uid` label will not reduce the current cardinality of the metric.

Reply:
There's no real way to reduce the cardinality of this metric while it still being useful IMO, but OTOH I think adding this by default is going to add a ton of load since it's 16 * 9 * # containers.

@dgrisonnet do you think a summary metric would scope that down enough? That at least brings us down to 3ish buckets instead of 16.

/hold

Reply:
I don't think that would help here, my main concern is that this metric has a cardinality of O(container) which can be problematic. Also, compared to the other container metrics exposed by the kubelet that also have a cardinality of O(container), this one will always expose information about containers that have been deleted. From the ingestion side, this is quite a huge difference.

I haven't really looked at the implementation details as I was not familiar with the codebase, but the cardinality should actually be O(probe) here, and I might not be as big of a problem as O(container)? My main concern was that O(container) can lead to cardinality explosion when they are a lot of short-lived containers created in a cluster, e.g CI jobs. But thinking about it again, I don't think it is relevant to add probes to this kind of container, so we might not see the theoretical cardinality spike that comes from O(container).

If we are sure that there are no use cases where short-lived containers use probes, I am fine with adding this metric. Otherwise, we might want to have it gated behind a flag and disabled by default.

Reply:
Thanks for the conversation here, folks. If anyone does have domain expertise in terms of the actual code, maybe a remark about what I'm doing wrong here? Test signal definitely suggests these changes are non-functional at this time.

PS I'll rebase now.

Reply:
As I mentioned in https://github.com/kubernetes/kubernetes/pull/104484#discussion_r718492374, removing `pod_uid` shouldn't have any impact on the cardinality since it is O(container) or even O(probe) depending on the implementation.

@ehashman correct me if I am wrong but I don't think there are any. We had a cardinality explosion in 1.22 where a change in a metric resulted in exposing 1.6m timeseries.

Reply:
it should help in situation of many short lived pods like was mentioned here: https://github.com/kubernetes/kubernetes/pull/104484#discussion_r743544075 In that case one pod will have a lot of pod_uid's.

Reply:
It will only barely help IMHO. Pods with variable names such as the ones created by replicasets will still be an issue. In this particular case, there should be as many `pod` as `pod_uid` and I don't think pods containing short-lived containers are necessarily stateful.

I would be fine with the following labels: "probe_type", "result", "container", "namespace". But I am still on the fence with `pod`. I can understand that it is an essential information here, but it is concerning in terms of cardinality.

A separate comment regarding `result`, I usually advise against adding error codes to durations metrics since in most cases you don't really care about the latency of requests based on the code that is returned. When setting up SLOs, you most likely want to know the percentage of requests that are failing, you'll get that with the counter metric and you want to know that the overall latency of your requests is under a certain threshold.

Reply:
removed "result" label

Reply:
Wow removed "result" and all the E2E tests pass! Can someone explain that to me?

Reply:
maybe other PRs between your tests fixed them. Serial for example is now passing

Reply:
😬

Reply:
I'm going to rebase now just so this has the latest bits when it is considered for merge

Reply:
> We already have the probe_total metric with these dimensions. What if we will have twice less buckets? (like 0.25, 0.5, 1, 2, 5, 10, 30, 60 or even 0.5, 1, 5, 10, 30)?

The buckets themselves aren't really the issue here since it basically adds a constant factor to the cardinality. Even though the `probe_total` metric has the same dimensions today, I don't think they are safe at all.

I am very much on the fence here since I can see the value of this metric and each one of its labels, but at the same time, the cardinality of O(container w/ probe) is quite frightening. 

I can see a few options moving forward:
- Go with the assumption that short-lived containers will not have probes. (It might bite us at some point)
- Introduce a mechanism to disable these high-cardinality metrics by default - doesn't exist yet
- Introduce a mechanism to bound the cardinality of these metrics - doesn't exist yet 

@ehashman @logicalhan - any thoughts on that? We could also bring that subject to the next sig meeting

Reply:
> Introduce a mechanism to bound the cardinality of these metrics - doesn't exist yet

Having pod names without deployment and replicaset randomness added for metrics would be great. Maybe also rely on some label if set to override the value of a pod dimension as a feature in the new k8s version? Including this metric today and implementing the name override feature in the next milestone may be a good idea.

> @ehashman @logicalhan - any thoughts on that? We could also bring that subject to the next sig meeting

@ehashman @logicalhan is there any opinion you can share today to include the metric as is into the 1.23 milestone?

Reply:
> Introduce a mechanism to disable these high-cardinality metrics by default - doesn't exist yet

I think moving this behind a feature gate makes plenty of sense to me, happy to do that. (If that requires a KEP happy to do so as well.)

Reply:
I'm torn on this. We have a mechanism to [disable high-cardinality metrics](https://kubernetes.io/docs/concepts/cluster-administration/system-metrics/#disabling-metrics), but it's opt out as opposed to opt in. A feature gate would be better from that perspective since it's opt-in as opposed to opt-out, but would require a KEP, and that would require a full enhancement cycle which seems like overkill for this.

Reply:
yeah same concerns as @ehashman on my side...